File: backend/src/com/origin/net/model/GameSession.java
package com.origin.net.model;

import com.origin.entity.Account;
import com.origin.net.WSServer;
import org.java_websocket.WebSocket;

public class GameSession
{
	private final WebSocket _connect;

	private final String _remoteAddr;

	private Account _account;

	public GameSession(WebSocket connect, String remoteAddr)
	{
		_connect = connect;
		_remoteAddr = remoteAddr;
	}

	public String getRemoteAddr()
	{
		return _remoteAddr;
	}

	public void send(String channel, Object data)
	{
		if (_connect != null && _connect.isOpen())
		{
			WSResponse response = new WSResponse();
			response.id = 0;
			response.data = data;
			response.channel = channel;

			_connect.send(WSServer.gsonSerialize.toJson(response));
		}
	}

	public void sendPing(String data)
	{
		if (_connect != null && _connect.isOpen())
		{
			_connect.send(data);
		}
	}

	public void setAccount(Account account)
	{
		_account = account;
	}

	public Account getAccount()
	{
		return _account;
	}
}

File: backend/src/com/origin/GameSession.kt
package com.origin

import com.google.gson.Gson
import com.origin.jooq.tables.records.AccountRecord
import com.origin.jooq.tables.records.CharacterRecord
import com.origin.model.Player
import com.origin.net.AuthorizeTokenResponse
import com.origin.net.GameRequestDTO
import com.origin.net.GameResponseDTO
import com.origin.net.ServerMessage
import io.ktor.websocket.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class GameSession(
    private val connect: DefaultWebSocketSession,
    private val token: String,
    val account: AccountRecord,
    val character: CharacterRecord,
) {
    companion object {
        val logger: Logger = LoggerFactory.getLogger(GameSession::class.java)

        /**
         * для сериализации пакетов (отправка клиенту)
         */
        private val gsonSerializer = Gson()
    }

    private lateinit var player: Player

    private var isDisconnected = false


    fun process(request: GameRequestDTO) {
        TODO("Not yet implemented")
    }

    suspend fun connected(request: GameRequestDTO) {
        ack(request, AuthorizeTokenResponse(character.id, ServerConfig.PROTO_VERSION))

        player = Player(character, this)
    }

    fun disconnected() {

    }

    /**
     * ответ на запрос клиента
     */
    private suspend inline fun ack(request: GameRequestDTO, d: Any? = null) {
        send(GameResponseDTO(request.id, d))
    }

    private suspend fun send(response: GameResponseDTO) {
        if (!isDisconnected) connect.outgoing.send(Frame.Text(gsonSerializer.toJson(response)))
    }

    suspend fun send(message: ServerMessage) {
        send(GameResponseDTO(message.channel, message))
    }

    suspend fun kick() {
        logger.warn("kick player")
        isDisconnected = true
        connect.close(CloseReason(CloseReason.Codes.NORMAL, "kicked"))
    }

    suspend fun logout() {
        isDisconnected = true
        connect.close(CloseReason(CloseReason.Codes.NORMAL, "logout"))
    }
}